### PR TITLE
Support more regional workspace cluster

### LINKED ISSUE #5596: Automatically choose workspace-cluster based on lowest latency.
Context: https://github.com/gitpod-io/gitpod/issues/5534#issuecomment-914967098

## Problem Statement
We currently have workspace clusters in one region in the EU and one region in the US. To offer service at a good latency (e.g. < 100ms), we will need more clusters, maybe as many as one or two per continent. See https://gcping.com/ for your personal latency to every google cloud region. See the [GCP network map](https://cloud.google.com/about/locations) for available regions and connections between them.

## Prior Art
* [Collect 'gcping' data from the dashboard](https://www.notion.so/gitpod/Collect-gcping-data-from-the-dashboard-dacb3eb6a0de4b809e4171a38d499e30) by @jankeromnes .

## Proposed Solution

The user's web browser should measure the latency for every available workspace cluster and send the measurements to the gitpod-server, so that the server can make an informed decision about what workspace-cluster is best for the user. 

## Considerations

* latency measurement should not slow down workspace startup time
* the decision what workspace-cluster to choose should remain with the gitpod-server, because in the future, other factors besides latency may influence the decision: Example: cluster health.

## Proposed Design Choices:

* to keep workspace startup fast, the latency measurement should be cached. For example in a cookie in the web-browser.
* to keep workspace startup fast, the latency measurement should preferable not be done when a workspace starts, but when a user visits any website of gitpod.
* every workspace clusters should have a public endpoint that can be "pinged" from the web browser for latency measurement.
* the server should make a cache-key and the ws-cluster-endpoints available to the users. The cache-key should encode the public IP address of the user, so that the latency will be measured again if the user changes his/her network.

## Example Flow 1:

1. the user visit gitpod.io/workspaces. 
2. the users browser receives `{'cache-key': 'FJJDSKD', "clusters": {"us07": "https://us07.gitpod.io/ping", "sing01": "https://sing01.gitpod.io/ping" } }`
3. the user browser measures the latency to all clusters in the background and stores the result in a cookie: `{"us07": 230, "sing01": 60}`
4. When the user opens a workspace, the cookie will be send to the gitpod-server and the server will use the latency measurement to chose the best workspace cluster.

## Example Flow 2: 
1. the user opens a workspace. The cookie is already there. No delay during workspace-start.

## Example Flow 3: 
1. the user opens a workspace. The cookie is not yet there. The is the case we want to avoid, but I don't think it can be avoided all the time.
2. measure the latency. Maybe the measurement can be aborted when the first workspace-cluster responds, because the first to respond will also be the one with the lowest latency (duh!). While there is the risk that the measurement is slightly inaccurate and repeated measurements would be needed for more accurate results, it seems like a good compromise to preserve fast workspace startup time. This way, if not cookie is present, 15 to ~200 ms will be added to to the workspace startup time.

### PR DESCRIPTION
**Pull request by csweichel**
## Description
Today workspace cluster choice is guided solely by the location of the application cluster the workspace is started from. This PR introduces an admission preference (introduced in #6164) which adds more regional support.

The current implementation in this branch is a rather rough first start. It does two things:
1. introduce the `region` admission preferences on workspace clusters. Each such admission preference carries a "round trip time endpoint" which is used to measure the RTT of a user to a particular region. We assume that if there are multiple clusters within the same "region" that their "RTT endpoints" yield equivalent results.
2. make the `start workspace` page measure the RTT and update the cluster preference for the user. This part I'm least happy with, because it's
  - super spammy and makes many more requests than it should
  - comes too late: only after the first workspace is running will this be done

It's particularly with point 2 that I'd need some support @gitpod-io/engineering-meta :)

## Related Issue(s)

Fixes #5596

## How to test
1. remove the default cluster registration by editing the `WSMAN_CFG` env var on server setting it to `W10K` (base64 for `[]`)
2. register a the ws-manager of the preview environment with `gpctl clusters register ...` make sure it's governed
3. register a ws-manager of another preview environment
4. use `gpctl clusters update admission-preference add ...` to point one cluster to a "slow" RTT endpoint, e.g. `ws-us25.gitpod.io`, and one to a "fast" RTT endpoint, e.g. `ws-eu25.gitpod.io` (depending on where in the world you're located)
5. start a workspace and observe the correct cluster choice

## Release Notes

```release-note
Support more regional workspace cluster
```

### CODE REVIEW

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **csweichel** and additionally assign **geropl** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @geropl` in a comment when ready.

Associated issue: *#6164*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**
- **[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**
- **[components/ws-manager-api/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager-api/OWNERS)**
- **[components/ws-manager-bridge/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager-bridge/OWNERS)**
- **[components/ws-manager-bridge-api/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager-bridge-api/OWNERS)**
- **[dev/gpctl/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/dev/gpctl/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/gitpod-protocol/src/protocol.ts [hunk 1]
```diff
@@ -1254,4 +1254,15 @@ export interface Terms {
 
 export interface WorkspaceClusterPreference {
     region?: string;
-}
+    // lrm == lastRegionMeasurement. Not writing the full name to save
+    // a few 100mb in the database.
+    lrm?: string;
```
Comment (jankeromnes) [inline components/gitpod-protocol/src/protocol.ts:1259]: Wait, for real? 🤯 (How was this calculated? Also, doesn't MySQL do compression? And sould we start using very short / unintelligible attribute names and values in every common shape? 😬 https://acronyms-suck.com/)

**Review by jankeromnes — COMMENTED** [review]
Exciting to see progress on this! ⚡ 👏 Many thanks for moving this forward. 🙏

Added a few drive-by comments in the dashboard part. Please disregard if unhelpful.

**Comment by JanKoehnlein** [comment]
My thoughts on this
1) Why not store the `lastMeasurement` date on the client (e.g. `localStorage`) instead of in the DB? It'd save one server call and allow to move this up in the component hierarchy as @jankeromnes recommends
2) Await on the measurement only if there is none yet. IMHO it's better to wait a few more seconds (AFAIKS its max 5s which could also be fine tuned) for the first workspace than being confronted with a horrible UX due to long latencies. We could show that to the user as a phase to raise acceptance if it takes longer (who would object a "Finding the best server for you")

**Comment by JanKoehnlein** [comment]
Code LGTM now, also squashed it. 

I admit that  I am stuck in the testing steps as I don't know how to register a different preview env's ws-manager: What is the URL? The one of the local preview env is a `dns://` URL which I assume won't work.

**Pull request closed without merging**